File: dijkstra.py
from dataclasses import dataclass, field
from typing import Dict, Set, List, Tuple, Optional
import heapq
import math

from grid import Net, Point

@dataclass
class Direction:
    x: int
    y: int

@dataclass
class DijkstraResult:
    start: Point
    directions: List[Direction]
    distance: float

@dataclass
class DijkstraModel:
    net: Net
    width: int
    height: int
    obstacles: Set[Point]
    # diagonal_obstacles: Set[Point]
    pending_net: Dict[Net, Set[Point]]
    start: Point
    end: Point

    def run(self) -> DijkstraResult:
        class State:
            def __init__(self, cost: float, position: Point):
                self.cost = cost
                self.position = position
            def __lt__(self, other):
                return self.cost < other.cost

        heap = [State(0.0, self.start)]
        dist = {self.start: 0.0}
        prev = {}

        cardinal_dirs = [(0, -1), (0, 1), (-1, 0), (1, 0)]
        diagonal_dirs = [(1, 1), (1, -1), (-1, 1), (-1, -1)]
        for key, points in self.pending_net.items():
            if key != self.net:
                self.obstacles.update(points)

        # print(self.obstacles)
        while heap:
            state = heapq.heappop(heap)
            cost, position = state.cost, state.position

            if position == self.end:
                break

            if dist.get(position, float('inf')) < cost:
                continue

            for dx, dy in cardinal_dirs:
                next_point = self.offset_point(position, dx, dy)
                if next_point and next_point not in self.obstacles:
                    new_cost = cost + 1.0
                    if new_cost < dist.get(next_point, float('inf')):
                        dist[next_point] = new_cost
                        prev[next_point] = position
                        heapq.heappush(heap, State(new_cost, next_point))

            for dx, dy in diagonal_dirs:
                next_point = self.offset_point(position, dx, dy)
                if next_point:
                    # top_left = self.offset_point(position, min(dx, 0), min(dy, 0))
                    if (next_point not in self.obstacles and 
                        (Point(position.x, next_point.y) not in self.obstacles or
                        Point(next_point.x, position.y) not in self.obstacles
                    )):
                        new_cost = cost + math.sqrt(2)
                        if new_cost < dist.get(next_point, float('inf')):
                            dist[next_point] = new_cost
                            prev[next_point] = position
                            heapq.heappush(heap, State(new_cost, next_point))

        directions = []
        current = self.end
        while current != self.start:
            if current not in prev:
                return DijkstraResult(self.start, [], float('inf'))
            prev_point = prev[current]
            if prev_point != self.start and prev_point != self.end:
                if self.net not in self.pending_net.keys():
                    self.pending_net[self.net] ={prev_point}
                elif prev_point not in self.pending_net[self.net]:
                    self.pending_net[self.net].add(prev_point)
            directions.append(Direction(current.x - prev_point.x, current.y - prev_point.y))
            current = prev_point

        directions.reverse()
        return DijkstraResult(self.start, directions, dist.get(self.end, float('inf')))

    def offset_point(self, point: Point, dx: int, dy: int) -> Optional[Point]:
        nx, ny = point.x + dx, point.y + dy
        if 0 <= nx < self.width and 0 <= ny < self.height:
            return Point(nx, ny)
        return None

File: gui_leptos_tauri/src-tauri/src/dijkstra.rs
use std::{cmp::Ordering, collections::{BTreeSet, BinaryHeap, HashMap, HashSet}};

use crate::{grid::Point, proba_grid::{Direction, TracePath}};

#[derive(Debug, Clone)]
pub struct DijkstraModel{
    pub width: usize,
    pub height: usize,
    pub obstacles: HashSet<Point>,
    pub diagonal_obstacles: HashSet<Point>, // obstacles that are diagonal traces
    pub start: Point,
    pub end: Point,
}

impl DijkstraModel {
    pub fn run(&self) -> Result<DijkstraResult, String> {
        let mut heap = BinaryHeap::new();
        let mut dist: HashMap<Point, f64> = HashMap::new();
        let mut prev: HashMap<Point, Point> = HashMap::new();

        #[derive(Debug, PartialEq)]
        struct State {
            cost: f64,
            position: Point,
        }

        impl Eq for State {}

        impl PartialOrd for State {
            fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
                other.cost.partial_cmp(&self.cost)
            }
        }

        impl Ord for State {
            fn cmp(&self, other: &Self) -> Ordering {
                self.partial_cmp(other).unwrap()
            }
        }

        heap.push(State { cost: 0.0, position: self.start });
        dist.insert(self.start, 0.0);

        let cardinal_dirs: [(i32, i32); 4] = [(0, -1), (0, 1), (-1, 0), (1, 0)];
        let diagonal_dirs: [(i32, i32); 4] = [(1, 1), (1, -1), (-1, 1), (-1, -1)];

        while let Some(State { cost, position }) = heap.pop() {
            if position == self.end {
                break;
            }
            // skip if we already found a shorter path
            if let Some(&d) = dist.get(&position) {
                if cost > d {
                    continue;
                }
            }
            // Cardinal moves
            for &(dx, dy) in &cardinal_dirs {
                if let Some(next) = self.offset_point(position, dx, dy) {
                    if self.obstacles.contains(&next) {
                        continue;
                    }
                    let next_cost = cost + 1.0;
                    if next_cost < *dist.get(&next).unwrap_or(&f64::INFINITY) {
                        dist.insert(next, next_cost);
                        prev.insert(next, position);
                        heap.push(State { cost: next_cost, position: next });
                    }
                }
            }
            // Diagonal moves
            for &(dx, dy) in &diagonal_dirs {
                if let Some(next) = self.offset_point(position, dx, dy) {
                    // top-left corner of the diagonal
                    let top_left = self.offset_point(position, dx.min(0), dy.min(0)).unwrap();
                    if self.obstacles.contains(&next) || self.diagonal_obstacles.contains(&top_left) {
                        continue;
                    }
                    let next_cost = cost + (2.0f64).sqrt();
                    if next_cost < *dist.get(&next).unwrap_or(&f64::INFINITY) {
                        dist.insert(next, next_cost);
                        prev.insert(next, position);
                        heap.push(State { cost: next_cost, position: next });
                    }
                }
            }
        }

        // Reconstruct path
        let mut trace_directions = Vec::new();
        let mut covered = BTreeSet::new();
        let mut diagonal_covered = BTreeSet::new();
        let mut current = self.end;
        covered.insert(current);
        while current != self.start {
            if let Some(&prev_point) = prev.get(&current) {
                let direction = Direction {
                    x: current.x as i32 - prev_point.x as i32,
                    y: current.y as i32 - prev_point.y as i32,
                };
                if direction.x != 0 && direction.y != 0 {
                    // if the direction is diagonal, we also add the diagonal trace
                    let diagonal_trace_point = Point {
                        x: prev_point.x.min(current.x),
                        y: prev_point.y.min(current.y),
                    };
                    diagonal_covered.insert(diagonal_trace_point);
                }
                trace_directions.push(direction);
                current = prev_point;
                covered.insert(current);
            } else {
                // No path found
                return Err("Dijkstra Algorithm Failed: No Path Found".to_string());
            }
        }
        trace_directions.reverse();
        let mut current = self.start;
        let trace_path: BTreeSet<Point> = trace_directions.iter().fold(BTreeSet::new(), |mut acc, dir| {
            current = Point {
                x: (current.x as i32 + dir.x) as usize,
                y: (current.y as i32 + dir.y) as usize,
            };
            acc.insert(current);
            acc
        });
        let trace_path = TracePath { 
            covered, 
            diagonal_covered
        };
        Ok(DijkstraResult {
            start: self.start,
            end: self.end,
            trace_path,
            trace_directions,
            distance: *dist.get(&self.end).unwrap_or(&f64::INFINITY),
        })
    }

    fn offset_point(&self, point: Point, dx: i32, dy: i32) -> Option<Point> {
        let nx = point.x as i32 + dx;
        let ny = point.y as i32 + dy;
        if nx >= 0 && ny >= 0 && (nx as usize) < self.width && (ny as usize) < self.height {
            Some(Point { x: nx as usize, y: ny as usize })
        } else {
            None
        }
    }
}

pub struct DijkstraResult{
    pub start: Point,
    pub end: Point, 
    pub trace_path: TracePath,
    pub trace_directions: Vec<Direction>,
    pub distance: f64,
}